pkg/datastore/postgres/repositories/header_sync_log_repository.go

```diff
@@ -0,0 +1,146 @@
+		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10) ON CONFLICT DO NOTHING`
+
+type HeaderSyncLogRepository struct {
+	db                *postgres.DB
+	addressRepository AddressRepository
+}
+
+func NewHeaderSyncLogRepository(db *postgres.DB) HeaderSyncLogRepository {
+	return HeaderSyncLogRepository{
+		db:                db,
+		addressRepository: AddressRepository{},
+	}
+}
+
+type headerSyncLog struct {
+	ID          int64
+	HeaderID    int64 `db:"header_id"`
+	Address     int64
+	Topics      pq.ByteaArray
+	Data        []byte
+	BlockNumber uint64 `db:"block_number"`
+	BlockHash   string `db:"block_hash"`
+	TxHash      string `db:"tx_hash"`
+	TxIndex     uint   `db:"tx_index"`
+	LogIndex    uint   `db:"log_index"`
+	Transformed bool
+	Raw         []byte
+}
+
+func (repository HeaderSyncLogRepository) GetUntransformedHeaderSyncLogs() ([]core.HeaderSyncLog, error) {
+	rows, queryErr := repository.db.Queryx(`SELECT * FROM public.header_sync_logs WHERE transformed = false`)
+	if queryErr != nil {
+		return nil, queryErr
+	}
+
+	var results []core.HeaderSyncLog
+	for rows.Next() {
+		var rawLog headerSyncLog
+		scanErr := rows.StructScan(&rawLog)
+		if scanErr != nil {
+			return nil, scanErr
+		}
+		var logTopics []common.Hash
+		for _, topic := range rawLog.Topics {
+			logTopics = append(logTopics, common.BytesToHash(topic))
+		}
+		address, addrErr := repository.addressRepository.GetAddressById(repository.db, rawLog.Address)
+		if addrErr != nil {
+			return nil, addrErr
+		}
+		reconstructedLog := types.Log{
+			Address:     common.HexToAddress(address),
+			Topics:      logTopics,
+			Data:        rawLog.Data,
+			BlockNumber: rawLog.BlockNumber,
+			TxHash:      common.HexToHash(rawLog.TxHash),
+			TxIndex:     rawLog.TxIndex,
+			BlockHash:   common.HexToHash(rawLog.BlockHash),
+			Index:       rawLog.LogIndex,
+			// TODO: revisit if not cascade deleting logs when header removed
+			// currently, fetched logs are cascade deleted if removed
+			Removed: false,
+		}
+		result := core.HeaderSyncLog{
+			ID:          rawLog.ID,
+			HeaderID:    rawLog.HeaderID,
+			Log:         reconstructedLog,
+			Transformed: rawLog.Transformed,
+		}
+		// TODO: Consider returning each result async to avoid keeping large result sets in memory
+		results = append(results, result)
+	}
+
+	return results, nil
+}
+
+func (repository HeaderSyncLogRepository) CreateHeaderSyncLogs(headerID int64, logs []types.Log) error {
+	tx, txErr := repository.db.Beginx()
+	if txErr != nil {
+		return txErr
+	}
+	for _, log := range logs {
+		err := repository.insertLog(headerID, log, tx)
+		if err != nil {
+			rollbackErr := tx.Rollback()
+			if rollbackErr != nil {
+				logrus.Errorf("failed to rollback header sync log insert: %s", rollbackErr.Error())
+			}
+			return err
+		}
+	}
+	return tx.Commit()
+}
+
+func (repository HeaderSyncLogRepository) insertLog(headerID int64, log types.Log, tx *sqlx.Tx) error {
+	topics := buildTopics(log)
```

Comment:
How come we need this? Are the topics just jumbled into the same byte array? No separator even?

Reply:
The topics column is an array of byte arrays. Felt nicer than having separate columns for each topic (especially given that many logs don't fill all four columns), but I'm open to reconsidering (especially if we need to do something like filter by topic0)

Reply:
Yah, makes sense